torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,731 @@
+        if placement.is_shard():
+            placement = cast(Shard, placement)
+            num_chunks = spec.mesh.size(dim=idx)
+            my_coord = spec.mesh.get_coordinate()[idx]
+            local_shard = placement._split_tensor(
+                local_shard, num_chunks, contiguous=False
+            )[0][my_coord]
+    return local_shard
+
+
+def partitioner(graph: GraphModule) -> GraphModule:
+    """
+    Graph partitioner that partitions the single device graph
+    to distributed graph
+    """
+
+    shape_adjustment_ops = [
+        aten.view.default,
+        aten.reshape.default,
+        aten.expand.default,
+    ]
+    # partition the graph to distributed
+    for node in graph.graph.nodes:
+        node_sharding = node.meta["sharding"]
+        # None sharding means this node don't need sharding
+        if node_sharding is None:
+            continue
+
+        if node.op == "placeholder":
+            # print(f">>>> processing placeholder: {node}")
+            out_spec = node_sharding.output_spec
+            if not hasattr(out_spec, "from_local"):
+                local_val = _to_local_shard(node.meta["val"], out_spec)
+                local_tensor_meta = _extract_tensor_metadata(local_val)
+                # update metadata
+                node.meta["val"] = local_val
+                node.meta["tensor_meta"] = local_tensor_meta
+        elif node.op == "call_function":
+            out_spec = node_sharding.output_spec
+            # for view related op that needs shape, adjust shape if needed
+            full_tensor = node.meta["val"]
+            full_shape = full_tensor.shape
+
+            # check if there's misaligned sharding, insert reshard if there is
+            expected_input_specs = node_sharding.input_specs
+            for idx, input_arg in enumerate(node.all_input_nodes):
+                input_arg_spec = input_arg.meta["sharding"].output_spec
+                input_arg_tensor = input_arg.meta["val"]
+                desired_spec = (
+                    out_spec
+                    if expected_input_specs is None
+                    else expected_input_specs[idx]
+                )
+                if input_arg_spec != desired_spec:
+                    # insert reshard operation
+                    def reshard_fn(local_tensor: torch.Tensor) -> torch.Tensor:
+                        return _redistribute_with_local_tensor(
+                            local_tensor,
+                            full_shape,
+                            out_spec.mesh,
+                            input_arg_spec.placements,
+                            desired_spec.placements,
+                        )
+
+                    reshard_gm = make_fx(reshard_fn)(input_arg_tensor)
+                    reshard_gm_nodes = list(reshard_gm.graph.nodes)
+                    input_node = reshard_gm_nodes[0]
+                    with graph.graph.inserting_before(node):
+                        output_node = graph.graph.graph_copy(
```

Comment:
looks like `graph_copy` will return a list of arg nodes. Do we need to unpack that?

Reply:
hmmm I didn't see any issue without unpacking, probably because if the copying gm have a single output it would return a single output instead.